fission/src/systems/physics/PhysicsSystem.ts

```diff
@@ -658,6 +661,91 @@ class PhysicsSystem extends WorldSystem {
+     */
+    private CreateConcaveShapeSettingsFromPart(
+        partDefinition: mirabuf.IPartDefinition
+    ): [Jolt.ShapeSettings, Jolt.Vec3, Jolt.Vec3] | undefined | null {
+        const settings = new JOLT.MeshShapeSettings()
+
+        settings.mMaxTrianglesPerLeaf = 8
+
+        settings.mTriangleVertices = new JOLT.VertexList()
+        settings.mIndexedTriangles = new JOLT.IndexedTriangleList()
+        settings.mMaterials = new JOLT.PhysicsMaterialList()
+
+        settings.mMaterials.push_back(new JOLT.PhysicsMaterial())
+
+        const min = new JOLT.Vec3(1000000.0, 1000000.0, 1000000.0)
+        const max = new JOLT.Vec3(-1000000.0, -1000000.0, -1000000.0)
+
+        partDefinition.bodies!.forEach(body => {
+            if (
+                body.triangleMesh &&
+                body.triangleMesh.mesh &&
+                body.triangleMesh.mesh.verts &&
+                body.triangleMesh.mesh.indices
+            ) {
+                const vertArr = body.triangleMesh.mesh.verts
+                for (let i = 0; i < body.triangleMesh.mesh.verts.length; i += 3) {
+                    const vert = MirabufFloatArr_JoltFloat3(vertArr, i)
+                    settings.mTriangleVertices.push_back(vert)
+                    this.UpdateMinMaxBounds(new JOLT.Vec3(vert), min, max)
+                    JOLT.destroy(vert)
+                }
+                const indexArr = body.triangleMesh.mesh.indices
+                for (let i = 0; i < body.triangleMesh.mesh.indices.length; i += 3) {
+                    settings.mIndexedTriangles.push_back(
+                        new JOLT.IndexedTriangle(indexArr.at(i)!, indexArr.at(i + 1)!, indexArr.at(i + 2)!, 0)
+                    )
+                }
+            }
+        })
+
+        if (settings.mTriangleVertices.size() < 4) {
+            JOLT.destroy(settings)
+            JOLT.destroy(min)
+            JOLT.destroy(max)
+            return
+        } else {
+            settings.Sanitize()
+            return [settings, min, max]
+        }
+    }
+
+    /**
+     * Raycast a ray into the physics scene.
+     *
+     * @param from Originating point of the ray
+     * @param dir Direction of the ray. Note: Length of dir specifies the maximum length it will check.
+     * @returns Either the hit results of the closest object in the ray's path, or undefined if nothing was hit.
+     */
+    public RayCast(from: Jolt.Vec3, dir: Jolt.Vec3): RayCastHit | undefined {
+        const ray = new JOLT.RayCast(from, dir)
+
+        const raySettings = new JOLT.RayCastSettings()
+        const collector = new JOLT.CastRayClosestHitCollisionCollector()
+        const bp_filter = new JOLT.BroadPhaseLayerFilter()
+        const object_filter = new JOLT.ObjectLayerFilter()
+        const body_filter = new JOLT.BodyFilter() // We don't want to filter out any bodies
+        const shape_filter = new JOLT.ShapeFilter() // We don't want to filter out any shapes
```

Comment:
If you don't mind me asking, why do we not want to filter these things?

Reply:
Because we want to cast the ray on everything and see what it hits